src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,224 @@
+{
+  pid_t pid;
+  while (pid = waitpid (-1, NULL, WNOHANG), pid > 0)
+    {
+      for (int i = 0; i < SOCKET_MODES; i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = -1;
+            }
+        }
+    }
+}
+
+int
+spawn_cockpit_ws (char *ws_path, int mode)
+{
+  int rv = 0;
+  switch (mode)
+    {
+    case WS_INSTANCE_HTTP:
+      rv = execl (ws_path, ws_path, "--no-tls", "--port", "0", NULL);
+      break;
+
+    case WS_INSTANCE_HTTP_REDIRECT:
+      rv = execl (ws_path, ws_path, "--proxy-tls-redirect",
+                  "--no-tls", "--port", "0", NULL);
+      break;
+
+    case WS_INSTANCE_HTTPS:
+      rv = execl (ws_path, ws_path, "--for-tls-proxy", "--port", "0", NULL);
+      break;
+
+    default:
+      perror ("Invalid mode");
+      rv = -1;
+    }
+  return rv;
+}
+
+int
+main (int argc, char *argv[])
+{
+  int rv = 0;
+
+  if (argc != 3)
+    {
+      perror ("Usage: socket-activation-helper $WS_PATH $SOCKETS_DIR");
+      goto cleanup;
+    }
+
+  int listen_fd[SOCKET_MODES];
+  struct sockaddr_un addr[SOCKET_MODES];
+  char *ws_path = argv[1];
+  char *https_socket_path = NULL;
+  char *http_redirect_socket_path = NULL;
+  char *http_socket_path = NULL;
+
+  rv = asprintf (&https_socket_path, "%s/%s", argv[2], HTTPS_SOCKET);
+  if (rv == -1)
+    goto cleanup;
+  rv =
+    asprintf (&http_redirect_socket_path, "%s/%s", argv[2],
+              HTTP_REDIRECT_SOCKET);
+  if (rv == -1)
+    goto cleanup;
+  rv = asprintf (&http_socket_path, "%s/%s", argv[2], HTTP_SOCKET);
+  if (rv == -1)
+    goto cleanup;
+
+  char *socket_paths[] =
+    { https_socket_path, http_redirect_socket_path, http_socket_path };
+
+  for (int i = 0; i < SOCKET_MODES; i++)
+    {
+      socket_to_pid[i] = -1;
+      unlink (socket_paths[i]);
+
+      /* create a listening socket for each cockpit-ws mode */
+      listen_fd[i] = socket (AF_UNIX, SOCK_STREAM, 0);
+
+      memset (&addr[i], 0, sizeof (struct sockaddr_un));
+      addr[i].sun_family = AF_UNIX;
+      strcpy (addr[i].sun_path, socket_paths[i]);
+      rv = bind (listen_fd[i],
+                 (struct sockaddr *) &addr[i], sizeof (addr[i]));
+      if (rv < 0)
+        {
+          perror ("bind() failed");
+          rv = -1;
+          goto cleanup;
+        }
+
+      rv = listen (listen_fd[i], 32);
+      if (rv < 0)
+        {
+          perror ("listen() failed");
+          rv = -1;
+          goto cleanup;
+        }
+
+      ufds[i].fd = listen_fd[i];
+      ufds[i].events = POLLIN;
+
+      if (fcntl (ufds[i].fd, F_SETFD, FD_CLOEXEC) == -1)
+        {
+          perror ("fcntl() failed");
+          rv = -1;
+          goto cleanup;
+        }
+    }
+
+  while (1)
+    {
+      /* Loop waiting for incoming connects on any of the connected sockets. */
+      rv = poll (ufds, SOCKET_MODES, -1);
+
+      if (rv == -1)
+        {
+          if (errno == EINTR)
+            continue;
```

Comment:
it's pretty common to write this patten like so:

```
do
  rv = syscall();
while (rv == -1 && errno == EINTR);
```

Matter of taste, though.  Yours is also fine.